radicale/storage.py

```diff
@@ -687,53 +862,87 @@ def delete(self, href=None):
+                self._meta = {}
+        return self._meta.get(key) if key else self._meta
 
     def set_meta(self, props):
-        if os.path.exists(self._props_path):
-            with open(self._props_path, encoding=self.encoding) as f:
-                old_props = json.load(f)
-                old_props.update(props)
-                props = old_props
-        props = {key: value for key, value in props.items() if value}
-        with self._atomic_write(self._props_path, "w+") as f:
-            json.dump(props, f)
+        new_props = self.get_meta()
+        new_props.update(props)
+        for key in tuple(new_props.keys()):
+            if not new_props[key]:
+                del new_props[key]
+        with self._atomic_write(self._props_path, "w") as f:
+            json.dump(new_props, f)
 
     @property
     def last_modified(self):
-        relevant_files = [self._filesystem_path] + [
-            path_to_filesystem(self._filesystem_path, href)
-            for href in self.list()]
-        if os.path.exists(self._props_path):
-            relevant_files.append(self._props_path)
+        relevant_files = chain(
+            (self._filesystem_path,),
+            (self._props_path,) if os.path.exists(self._props_path) else (),
+            map(lambda x: os.path.join(self._filesystem_path, x), self.list()))
         last = max(map(os.path.getmtime, relevant_files))
         return time.strftime("%a, %d %b %Y %H:%M:%S GMT", time.gmtime(last))
 
     def serialize(self):
-        items = []
-        time_begin = datetime.datetime.now()
-        for href in self.list():
-            items.append(self.get(href).item)
-        time_end = datetime.datetime.now()
-        self.logger.info(
-            "Collection read %d items in %s sec from %s", len(items),
-            (time_end - time_begin).total_seconds(), self._filesystem_path)
+        # serialize collection
         if self.get_meta("tag") == "VCALENDAR":
-            collection = vobject.iCalendar()
-            for item in items:
-                for content in ("vevent", "vtodo", "vjournal"):
-                    if content in item.contents:
-                        for item_part in getattr(item, "%s_list" % content):
-                            collection.add(item_part)
-                        break
-            return collection.serialize()
+            vtimezones = ""
+            included_tzids = set()
+            vtimezone = []
+            tzid = None
+            components = ""
+            for item in self.get_all():
+                depth = 0
+                for line in item.serialize().split("\r\n"):
+                    if line.startswith("BEGIN:"):
+                        depth += 1
+                    if depth == 2 and line == "BEGIN:VTIMEZONE":
+                        vtimezone.append(line)
+                    elif vtimezone:
+                        vtimezone.append(line)
+                        if depth == 2 and line.startswith("TZID:"):
```

Comment:
I mean this case (which *may* be an invalid vevent, but this *may* change in some revision of the standard)
```
BEGIN:Rubbish
BEGIN:VTIMEZONE
TZID:
…
```